File: backend/python/app/services/implementations/entity_service.py
# postgresql {
from ...models.entity import Entity
from ...models import db
from ..interfaces.entity_service import IEntityService
# file-storage {
from ..interfaces.file_storage_service import IFileStorageService
from uuid import uuid4
# } file-storage


class EntityService(IEntityService):
    # file-storage {
    def __init__(self, logger, file_storage_service: IFileStorageService):
    # } file-storage
    # no-file-storage {
    def __init__(self, logger):
    # } no-file-storage
        self.logger = logger
        # file-storage {
        self.file_storage_service = file_storage_service
        # } file-storage

    def get_entities(self):
        # Entity is a SQLAlchemy model, we can use convenient methods provided
        # by SQLAlchemy like query.all() to query the data
        return [result.to_dict() for result in Entity.query.all()]

    def get_entity(self, id):
        # get queries by the primary key, which is id for the Entity table
        entity = Entity.query.get(id)
        if entity is None:
            self.logger.error("Invalid id")
            raise Exception("Invalid id")
        return entity.to_dict()

    def create_entity(self, entity):
        try:
            # file-storage {
            file_name = None
            if entity.file:
                file_name = str(uuid4())
                self.file_storage_service.create_file(
                    file_name, entity.file, entity.file.content_type
                )
            entity.__dict__.pop("file", None)
            new_entity = Entity(**entity.__dict__, file_name=file_name)
            # } file-storage
            # no-file-storage {
            new_entity = Entity(**entity.__dict__)
            # } no-file-storage
        except Exception as error:
            self.logger.error(str(error))
            raise error

        db.session.add(new_entity)
        # remember to commit to actually persist into the database
        db.session.commit()

        return new_entity.to_dict()

    def update_entity(self, id, entity):
        # file-storage {
        current_entity = Entity.query.get(id)

        if current_entity is None:
            self.logger.error("Invalid id")
            raise Exception("Invalid id")

        file_name = current_entity.file_name
        if entity.file:
            if file_name:
                self.file_storage_service.update_file(
                    file_name, entity.file, entity.file.content_type
                )
            else:
                file_name = str(uuid4())
                self.file_storage_service.create_file(
                    file_name, entity.file, entity.file.content_type
                )
        elif file_name:
            self.file_storage_service.delete_file(file_name)
            file_name = None

        entity_dict = entity.__dict__
        entity_dict.pop("file", None)
        entity_dict.update({"file_name": file_name})

        Entity.query.filter_by(id=id).update(entity_dict)
        # } file-storage
        # no-file-storage {
        Entity.query.filter_by(id=id).update(entity.__dict__)
        # } no-file-storage
        updated_entity = Entity.query.get(id)
        db.session.commit()

        if updated_entity is None:
            self.logger.error("Invalid id")
            raise Exception("Invalid id")
        return updated_entity.to_dict()

    def delete_entity(self, id):
        # file-storage {
        deleted_entity = Entity.query.get(id)
        # } file-storage
        deleted = Entity.query.filter_by(id=id).delete()
        db.session.commit()

        # deleted is the number of rows deleted
        if deleted == 1:
            # file-storage {
            if deleted_entity.file_name:
                self.file_storage_service.delete_file(deleted_entity.file_name)
            # } file-storage
            return id

        self.logger.error("Invalid id")
        raise Exception("Invalid id")

# } postgresql
# mongodb {
# file-storage {
from ..interfaces.file_storage_service import IFileStorageService
# } file-storage
from ...models.entity import Entity
from ..interfaces.entity_service import IEntityService
# file-storage {
from uuid import uuid4
# } file-storage


class EntityService(IEntityService):
    # file-storage {
    def __init__(self, logger, file_storage_service: IFileStorageService):
    # } file-storage
    # no-file-storage {
    def __init__(self, logger):
    # } no-file-storage
        self.logger = logger
        # file-storage {
        self.file_storage_service = file_storage_service
        # } file-storage

    def get_entities(self):
        entities = []

        for result in Entity.objects:
            result_dict = result.to_serializable_dict()
            entities.append(result_dict)

        return entities

    def get_entity(self, id):
        entity = Entity.objects.get(id=id)

        if entity is None:
            self.logger.error("Invalid id")
            raise Exception("Invalid id")

        return entity.to_serializable_dict()

    def create_entity(self, entity):
        try:
            # file-storage {
            file_name = None
            if entity.file:
                file_name = str(uuid4())
                self.file_storage_service.create_file(
                    file_name, entity.file, entity.file.content_type
                )
            entity.__dict__.pop("file", None)
            new_entity = Entity(**entity.__dict__, file_name=file_name)
            # } file-storage
            # no-file-storage {
            new_entity = Entity(**entity.__dict__)
            # } no-file-storage
            new_entity.save()
        except Exception as error:
            self.logger.error(str(error))
            raise error

        return new_entity.to_serializable_dict()

    def update_entity(self, id, entity):
        # file-storage {
        current_entity = Entity.objects.get(id=id)
        if current_entity is None:
            self.logger.error("Invalid id")
            raise Exception("Invalid id")
        file_name = current_entity.file_name
        if entity.file:
            if file_name:
                self.file_storage_service.update_file(
                    file_name, entity.file, entity.file.content_type
                )
            else:
                file_name = str(uuid4())
                self.file_storage_service.create_file(
                    file_name, entity.file, entity.file.content_type
                )
        elif file_name:
            self.file_storage_service.delete_file(file_name)
            file_name = None

        entity.__dict__.pop("file", None)
        updated_entity = Entity.objects(id=id).modify(
            new=True, **entity.__dict__, file_name=file_name
        )
        # } file-storage
        # no-file-storage {
        updated_entity = Entity.objects(id=id).modify(new=True, **entity.__dict__)
        # } no-file-storage

        if updated_entity is None:
            self.logger.error("Invalid id")
            raise Exception("Invalid id")

        return updated_entity.to_serializable_dict()

    def delete_entity(self, id):
        try:
            # file-storage {
            file_name = Entity.objects.get(id=id).file_name
            # } file-storage
            deleted = Entity.objects(id=id).modify(remove=True)
            # file-storage {
            if deleted:
                self.file_storage_service.delete_file(file_name)
            # } file-storage
            return id
        except Exception as error:
            self.logger.error(str(error))
            raise error

# } mongodb

File: backend/typescript/services/implementations/entityService.ts
// mongodb {
import { v4 as uuidv4 } from "uuid";

import MgEntity, { Entity } from "../../models/entity.model";
import {
  IEntityService,
  EntityRequestDTO,
  EntityResponseDTO,
} from "../interfaces/IEntityService";
import IFileStorageService from "../interfaces/fileStorageService";
import { getErrorMessage } from "../../utilities/errorUtils";
import logger from "../../utilities/logger";

const Logger = logger(__filename);

class EntityService implements IEntityService {
  storageService: IFileStorageService;

  constructor(storageService: IFileStorageService) {
    this.storageService = storageService;
  }

  /* eslint-disable class-methods-use-this */
  async getEntity(id: string): Promise<EntityResponseDTO> {
    let entity: Entity | null;
    try {
      entity = await MgEntity.findById(id);
      if (!entity) {
        throw new Error(`Entity id ${id} not found`);
      }
    } catch (error: unknown) {
      Logger.error(`Failed to get entity. Reason = ${getErrorMessage(error)}`);
      throw error;
    }

    return {
      id: entity.id,
      stringField: entity.stringField,
      intField: entity.intField,
      enumField: entity.enumField,
      stringArrayField: entity.stringArrayField,
      boolField: entity.boolField,
      fileName: entity.fileName,
    };
  }

  async getEntities(): Promise<EntityResponseDTO[]> {
    try {
      const entities: Array<Entity> = await MgEntity.find();
      return entities.map((entity) => ({
        id: entity.id,
        stringField: entity.stringField,
        intField: entity.intField,
        enumField: entity.enumField,
        stringArrayField: entity.stringArrayField,
        boolField: entity.boolField,
        fileName: entity.fileName,
      }));
    } catch (error: unknown) {
      Logger.error(
        `Failed to get entities. Reason = ${getErrorMessage(error)}`,
      );
      throw error;
    }
  }

  async createEntity(entity: EntityRequestDTO): Promise<EntityResponseDTO> {
    let newEntity: Entity | null;
    const fileName = entity.filePath ? uuidv4() : "";
    try {
      if (entity.filePath) {
        await this.storageService.createFile(
          fileName,
          entity.filePath,
          entity.fileContentType,
        );
      }
      newEntity = await MgEntity.create({ ...entity, fileName });
    } catch (error: unknown) {
      Logger.error(
        `Failed to create entity. Reason = ${getErrorMessage(error)}`,
      );
      throw error;
    }
    return {
      id: newEntity.id,
      stringField: newEntity.stringField,
      intField: newEntity.intField,
      enumField: newEntity.enumField,
      stringArrayField: newEntity.stringArrayField,
      boolField: newEntity.boolField,
      fileName,
    };
  }

  async updateEntity(
    id: string,
    entity: EntityRequestDTO,
  ): Promise<EntityResponseDTO | null> {
    let updatedEntity: Entity | null;
    let fileName = "";
    try {
      const currentEntity = await MgEntity.findById(id, "fileName");
      const currentFileName = currentEntity?.fileName;
      if (entity.filePath) {
        fileName = currentFileName || uuidv4();
        if (currentFileName) {
          await this.storageService.updateFile(
            fileName,
            entity.filePath,
            entity.fileContentType,
          );
        } else {
          await this.storageService.createFile(
            fileName,
            entity.filePath,
            entity.fileContentType,
          );
        }
      } else if (currentFileName) {
        await this.storageService.deleteFile(currentFileName);
      }
      updatedEntity = await MgEntity.findByIdAndUpdate(
        id,
        { ...entity, fileName },
        {
          new: true,
          runValidators: true,
        },
      );
      if (!updatedEntity) {
        throw new Error(`Entity id ${id} not found`);
      }
    } catch (error: unknown) {
      Logger.error(
        `Failed to update entity. Reason = ${getErrorMessage(error)}`,
      );
      throw error;
    }
    return {
      id: updatedEntity.id,
      stringField: updatedEntity.stringField,
      intField: updatedEntity.intField,
      enumField: updatedEntity.enumField,
      stringArrayField: updatedEntity.stringArrayField,
      boolField: updatedEntity.boolField,
      fileName,
    };
  }

  async deleteEntity(id: string): Promise<string> {
    try {
      const deletedEntity: Entity | null = await MgEntity.findByIdAndDelete(id);
      if (!deletedEntity) {
        throw new Error(`Entity id ${id} not found`);
      }
      if (deletedEntity.fileName) {
        await this.storageService.deleteFile(deletedEntity.fileName);
      }
      return id;
    } catch (error: unknown) {
      Logger.error(
        `Failed to delete entity. Reason = ${getErrorMessage(error)}`,
      );
      throw error;
    }
  }
}

export default EntityService;

// } mongodb
// postgresql {
import { v4 as uuidv4 } from "uuid";

import { PrismaClient, entities, enum_entities_enum_field} from '@prisma/client';

const prisma = new PrismaClient();

import {
  IEntityService,
  EntityRequestDTO,
  EntityResponseDTO,
} from "../interfaces/IEntityService";
import IFileStorageService from "../interfaces/fileStorageService";
import { getErrorMessage } from "../../utilities/errorUtils";
import logger from "../../utilities/logger";
import { EnumType } from "typescript";

const Logger = logger(__filename);

class EntityService implements IEntityService {
  storageService: IFileStorageService;

  constructor(storageService: IFileStorageService) {
    this.storageService = storageService;
  }

  /* eslint-disable class-methods-use-this */
  async getEntity(id: string): Promise<EntityResponseDTO> {
    let entity: entities | null;
    try {
      entity = await prisma.entities.findUnique({ where: { id: Number(id) } });
      if (!entity) {
        throw new Error(`Entity id ${id} not found`);
      }
    } catch (error: unknown) {
      Logger.error(`Failed to get entity. Reason = ${getErrorMessage(error)}`);
      throw error;
    }

    return {
      id: String(entity.id),
      stringField: entity.string_field,
      intField: entity.int_field,
      enumField: entity.enum_field as enum_entities_enum_field,
      stringArrayField: entity.string_array_field,
      boolField: entity.bool_field,
      fileName: entity.file_name,
    };
  }

  async getEntities(): Promise<EntityResponseDTO[]> {
    try {
      const entities: Array<entities> = await prisma.entities.findMany();
      return entities.map((entity) => ({
        id: String(entity.id),
        stringField: entity.string_field,
        intField: entity.int_field,
        enumField: entity.enum_field as enum_entities_enum_field,
        stringArrayField: entity.string_array_field,
        boolField: entity.bool_field,
        fileName: entity.file_name,
      }));
    } catch (error: unknown) {
      Logger.error(
        `Failed to get entities. Reason = ${getErrorMessage(error)}`,
      );
      throw error;
    }
  }

  async createEntity(entity: EntityRequestDTO): Promise<EntityResponseDTO> {
    let newEntity: entities | null;
    const fileName = entity.filePath ? uuidv4() : "";
    try {
      if (entity.filePath) {
        await this.storageService.createFile(
          fileName,
          entity.filePath,
          entity.fileContentType,
        );
      }
      newEntity = await prisma.entities.create({
        data: {
          string_field: entity.stringField,
          int_field: entity.intField,
          enum_field: entity.enumField as enum_entities_enum_field,
          string_array_field: entity.stringArrayField,
          bool_field: entity.boolField,
          file_name: fileName,
        },
      });
    } catch (error: unknown) {
      Logger.error(
        `Failed to create entity. Reason = ${getErrorMessage(error)}`,
      );
      throw error;
    }
    return {
      id: String(newEntity.id),
      stringField: newEntity.string_field,
      intField: newEntity.int_field,
      enumField: newEntity.enum_field as enum_entities_enum_field,
      stringArrayField: newEntity.string_array_field,
      boolField: newEntity.bool_field,
      fileName,
    };
  }

  async updateEntity(
    id: string,
    entity: EntityRequestDTO,
  ): Promise<EntityResponseDTO | null> {
    let resultingEntity: entities | null;
    let fileName = "";
    try {
      const currentEntity = await prisma.entities.findUnique({
        where: { id: Number(id) },
        select: {
          file_name: true,
        }
      }) as { file_name: string } | null ;
      const currentFileName = currentEntity?.file_name;
      if (entity.filePath) {
        fileName = currentFileName || uuidv4();
        if (currentFileName) {
          await this.storageService.updateFile(
            fileName,
            entity.filePath,
            entity.fileContentType,
          );
        } else {
          await this.storageService.createFile(
            fileName,
            entity.filePath,
            entity.fileContentType,
          );
        }
      } else if (currentFileName) {
        await this.storageService.deleteFile(currentFileName);
      }
      resultingEntity = await prisma.entities.update({
        where: { id: Number(id) },
        data :
        {
          string_field: entity.stringField,
          int_field: entity.intField,
          enum_field: entity.enumField as enum_entities_enum_field,
          string_array_field: entity.stringArrayField,
          bool_field: entity.boolField,
          file_name: fileName,
        },
      });
    } catch (error: unknown) {
      Logger.error(
        `Failed to update entity. Reason = ${getErrorMessage(error)}`,
      );
      throw error;
    }
    if (!resultingEntity) {
      throw new Error(`Failed to update entity with id ${id}`);
    }
    return {
      id: String(resultingEntity.id),
      stringField: resultingEntity.string_field,
      intField: resultingEntity.int_field,
      enumField: resultingEntity.enum_field as enum_entities_enum_field,
      stringArrayField: resultingEntity.string_array_field,
      boolField: resultingEntity.bool_field,
      fileName,
    };
  }

  async deleteEntity(id: string): Promise<string> {
    try {
      const entityToDelete = await prisma.entities.findUnique({ where: { id: Number(id) } });
      await prisma.entities.delete({ where: { id: Number(id) } });

      if (!entityToDelete) {
        throw new Error(`Entity id ${id} not found`);
      }
      if (entityToDelete.file_name) {
        await this.storageService.deleteFile(entityToDelete.file_name);
      }
      return id;
    } catch (error: unknown) {
      Logger.error(
        `Failed to delete entity. Reason = ${getErrorMessage(error)}`,
      );
      throw error;
    }
  }
}

export default EntityService;
// } postgresql